### DISCUSSION TITLE
week09 QUIZ

### DISCUSSION THREAD

**Discussion by hunteac** [damdam6/ModernJavascript-Study#30] [Q&A]
1. "함수의 상위 스코프를 결정한다."를 렉시컬 환경의 관점으로 설명하세요.

2. 다음 'multiply' 함수를 구현하세요.

```javascript
multiply(2, 4); // 8
multiply(3, 5); // 15

const double = multiply(2);
double(2); // 2
double(8); // 16

const hexa = multiply(6);
hexa(6); // 36
hexa(10); // 60
```

3. 아래의 코드는 잘 동작하지만 오류를 발생시킬 가능성이 있다. 클로저를 활용해서 안전한 코드를 작성하세요.

```javascript
let num = 0;

const increase = function () {
  return ++num;
};

console.log(increase());
console.log(increase());
console.log(increase());
```

4. 클래스 내부에 정의된 정적 메서드와 프로토타입 메서드의 차이를 설명하세요.

5. 아래의 코드에 존재하는 name을 private 필드로 변경하세요.

```javascript
class Person {
  name = '';
}

// 인스턴스 생성
const me = new Person();
```

**Comment by sgryu23** [comment]
1. "함수의 상위 스코프를 결정한다."를 렉시컬 환경의 관점으로 설명하세요.
- 자바스크립트 엔진은 함수를 어디서 호출했는지가 아니라 어디에 정의했는지에 따라 함수의 상위 스코프를 결정한다.

2. 다음 'multiply' 함수를 구현하세요.
```js
function multiply(...args) {
  // 인자가 두 개 전달되면 두 값의 곱셈 결과를 반환
  if (args.length === 2) {
    return args[0] * args[1];
  }
  // 인자가 하나만 전달되는 경우 고차 함수를 반환
  else if (args.length === 1) {
    return function(secondArg) {
      return args[0] * secondArg;
    }
  }
}
```

3. 아래의 코드는 잘 동작하지만 오류를 발생시킬 가능성이 있다. 클로저를 활용해서 안전한 코드를 작성하세요.
```js
function createCounter() {
  let num = 0;
    
  return function() {
    return ++num;
  };
}

const increaseByOne = createCounter();

console.log(increaseByOne());  // 1
console.log(increaseByOne());  // 2
console.log(increaseByOne());  // 3
```

4. 클래스 내부에 정의된 정적 메서드와 프로토타입 메서드의 차이를 설명하세요.
#### (1) 호출 방식 차이
  - 정적 메서드는 클래스 자체에 바인딩이 됩니다. 인스턴스를 생성하지 않고도 클래스 이름을 사용하여 호출할 수 있습니다.
  - 프로토타입 메서드는 클래스의 인스턴스를 통해 호출됩니다. 이 메서드들은 인스턴스의 프로토타입 체인에 추가되므로 모든 메서드를 상속받아 사용할 수 있습니다.

#### (2) 상속의 차이
  - 정적 메서드는 클래스 상속 시에만 상속됩니다.
  - 프로토타입 메서드는 모든 인스턴스에 상속됩니다. 하위 클래스에서 생성된 인스턴스는 상위 클래스의 프로토타입 메서드에 접근할 수 있습니다.


5. 아래의 코드에 존재하는 name을 private 필드로 변경하세요.
```js
class Person {
  // '#' prefix를 붙여 private로 변경할 수 있다.
  #name = '';

  constructor(name = '') {
    this.#name = name;
  }
}

// 인스턴스 생성
const me = new Person('Lee');
console.log(me.#name);  // SyntaxError: Private field '#name' must be declared in an enclosing class
```

**Comment by damdam6** [comment]
1. "함수의 상위 스코프를 결정한다."를 렉시컬 환경의 관점으로 설명하세요.

렉시컬 환경(Lexical Environment)은 코드가 작성된 시점의 환경을 기반으로 상위 스코프를 결정한다. 함수를 어디서 호출하든지 간에 함수의 렉시컬 환경은 함수가 정의된 위치에 따라 이미 결정되어 있다.

2. 다음 'multiply' 함수를 구현하세요.

```
function multiply(a, b) {
    if (b === undefined) {
        return function(b) {
            return a * b;
        };
    }
    else {
        return a * b;
    }
}

```

3. 아래의 코드는 잘 동작하지만 오류를 발생시킬 가능성이 있다. 클로저를 활용해서 안전한 코드를 작성하세요.
```
const increase = (function () {
  let num = 0;
  return function () {
    return ++num;
  };
})();
```

4. 클래스 내부에 정의된 정적 메서드와 프로토타입 메서드의 차이를 설명하세요.

**정적 메서드(Static Method)**
-  클래스 자체에 바인딩
- 인스턴스를 생성하지 않고도 호출 가능
- 'this'없다.

**프로토타입 메서드(Prototype Method)**
- 인스턴스에 바인딩 되어 있음.
- 인스턴스 호출로 사용가능
- 메서드는 인스턴스끼리 공유

5. 아래의 코드에 존재하는 name을 private 필드로 변경하세요.

```

class Person {
  #name = '';
  
    getName() {
    return this.#name;
  }

  setName(name) {
    this.#name = name;
  }
}

// 인스턴스 생성
const me = new Person();

```
